File: Senseair-K30_arduino/K30.cpp
#include "K30.h"

// Constructor to initialize the I2C communication with the specified SDA and SCL pins
K30::K30(uint8_t scl_pin, uint8_t sda_pin, uint8_t co2_addr) {
  Wire.begin(sda_pin, scl_pin); // Initialize I2C with specified SDA and SCL pins
  this->co2_addr = co2_addr;
}

// Method to calculate checksum from the buffer (byte[0] + byte[1] + byte[2])
void K30::_calculate_checksum(byte* buffer) {
  return (buffer[0] + buffer[1] + buffer[2]) & 0xFF;
}

// Method to read CO2 data from the sensor
uint16_t K30::read_co2() {
  // Write request to sensor
  Wire.beginTransmission(co2_addr);
  Wire.write(0x22); // Start byte
  Wire.write(0x00);
  Wire.write(0x08);
  Wire.write(0x2A);
  if (Wire.endTransmission() != 0) {
    Serial.println("I2C write error");
    return 0;
  }

  // Wait for the sensor to process
  delay(10); // wait 10ms

  // Read response from sensor
  Wire.requestFrom(co2_addr, 4);
  if (Wire.available() == 4) {
    byte buffer[4];
    for (int i = 0; i < 4; i++) {
      buffer[i] = Wire.read();
    }

    // Validate checksum
    uint16_t co2_value = (buffer[1] << 8) | buffer[2];
    byte checksum = _calculate_checksum(buffer);
    if (checksum == buffer[3]) {
      return co2_value;
    } else {
      Serial.println("Checksum validation failed");
      return 0;
    }
  } else {
    Serial.println("I2C read error");
    return 0;
  }
}

File: Senseair-K30_Micropython/K30.py
from machine import Pin, I2C
import time

class K30CO2Sensor:
    def __init__(self, scl_pin, sda_pin, freq=100000, co2_addr=0x68):
        self.i2c = I2C(0, scl=Pin(scl_pin), sda=Pin(sda_pin), freq=freq)
        self.co2_addr = co2_addr

    def _calculate_checksum(self, buffer):
        return sum(buffer[0:3]) & 0xFF

    def read_co2(self):
        # Write request to sensor
        try:
            self.i2c.writeto(self.co2_addr, bytearray([0x22, 0x00, 0x08, 0x2A]))
        except OSError as e:
            print("I2C write error:", e)
            return None

        # Wait for the sensor to process
        time.sleep(0.01)

        # Read response from sensor
        try:
            buffer = self.i2c.readfrom(self.co2_addr, 4)
        except OSError as e:
            print("I2C read error:", e)
            return None

        # Validate checksum
        co2_value = (buffer[1] << 8) | buffer[2]
        checksum = self._calculate_checksum(buffer)
        if checksum == buffer[3]:
            return co2_value
        else:
            print("Checksum validation failed")
            return None

